# Анализ данных на Python

*Алла Тамбовцева*

##  Введение в датафреймы Pandas: загрузка и описание данных

### Загрузка данных

Как мы уже убедились, библиотека `pandas` позволяет создавать датафреймы – удобные структуры для данных в табличном виде. Датафрейм можно получить из списка списков, из списка словарей, из словаря, простого или сложного. Но если наша задача – не преобразовать собранные данные (например, выгруженные с веб-страницы), а обработать уже готовые из имеющегося файла, библиотека `pandas` тоже будет чрезвычайно полезна. Импортируем её с сокращённым названием `pd`:

In [1]:
import pandas as pd

Библиотека pandas умеет загружать данные из файлов разных форматов. Среди них файлы Excel (расширения `.xls` и `.xlsx`), Stata (расширение `.dta`), SPSS (расширение `.sav`), текстовые файлы в формате TXT, CSV, JSON и другие. Мы поработаем с форматом CSV, одним из самых распространенных машиночитаемых форматов, название которого расшифровывается как *Comma Separated Values*, то есть «значения, разделённые запятыми».

На практике можно столкнуться и с менее классическими вариантами этого формата. Например, если открыть файл CSV в Excel, он автоматически изменит разделитель `,` на `;`, чтобы точно не перепутать запятую как разделитель столбцов и запятую как десятичный разделить (между целой и дробной частью в числах). В таких случаях символ для разделителя нужно будет указать внутри аргумента `sep` в соответствующей функции, например, `sep = ";"`.

Перейдём к нашему файлу `Salaries.csv`. Если файл сохранён на компьютере, у нас два пути:

* Поместить его в рабочую папку (рядом с ipynb-файлом, в котором пишем код для загрузки данных) через кнопку *Upload* на главной странице *Home* в Jupyter.
* Найти файл на компьютере, кликнуть правой клавишей, зайти в свойства и скопировать полный путь к файлу из его расположения.

В первом случае все просто, вызываем функцию `read_csv()` и помещаем в неё название файла в виде строки:

In [3]:
df = pd.read_csv("Salaries.csv")

Во втором случае нужно учесть важный момент: слэши в пути должны быть такие `/`,  а не такие `\`. Слэш в обратную сторону внутри строки Python воспринимает как специальный символ, плюс, сочетание `\U`, которое неизбежно возникает при наличии папки `Users` будет намекать Python на кодировку символов в Unicode. На Mac и Linux таких слэшей в пути к файлу не возникает, а на Windows, чтобы не заменять каждый слэш вручную, можно перед кавычками поставить букву `r`:

In [ ]:
df = pd.read_csv(r"C:\Users\student\Desktop\Salaries.csv")

*Примечание для Google Colab.* Загрузить файл с данными в облачное хранилище можно через кнопку *Files* (значок папки слева от рабочей области с ячейками), при нажатии на которую появляется возможность выбрать файл с компьютера (значок стрелки). После добавления файла его можно выбрать, кликнуть на три точки справа от названия, скопировать путь через *Copy path* и вставить его в функцию `read_csv()`.

Итак, если никаких сообщений об ошибках не было (если `FileNotFoundError`, файл не найден, проверьте, нет ли опечаток в названии, и лежит ли файл в той папке, к которой мы прописали путь или в которой лежит текущий ipynb-файл), данные благополучно сохранились в виде датафрейма `df`.

*Примечание.* Если вы планируете работать с файлами Excel, логика та же, только функция будет `read_excel()`. По умолчанию она считывает только первый лист файла, но в аргументе `sheet_name` можно указать индекс или название листа. Про другие, более продвинутые, опции вроде тех, что позволяют пропускать первые несколько строк в файле или, наоборот, загружать только определенный фрагмент таблицы, можно узнать, запросив `help(pd.read_excel)`.

### Знакомство с данными

В этом файле сохранены данные по сотрудникам университета в США, а именно следующие их характеристики:

* `rank`: должность;
* `discipline`: тип преподаваемой дисциплины (A – теоретическая, B – практическая);
* `yrs.since.phd`: число лет с момента получения степени PhD;
* `yrs.service`: число лет опыта работы;
* `sex`: пол;
* `salary`: заработная плата за 9 месяцев, в долларах.

Приступим к изучению содержимого датафрейма.

Если датафрейм большой, а мы хотим быстро посмотреть на то, какого он вида, можем запросить на экран только первые строчки или последние:

In [4]:
df.head()

Unnamed: 0      rank discipline  yrs.since.phd  yrs.service   sex  salary
0           1      Prof          B             19           18  Male  139750
1           2      Prof          B             20           16  Male  173200
2           3  AsstProf          B              4            3  Male   79750
3           4      Prof          B             45           39  Male  115000
4           5      Prof          B             40           41  Male  141500

In [5]:
df.tail()

Unnamed: 0      rank discipline  yrs.since.phd  yrs.service   sex  salary
392         393      Prof          A             33           30  Male  103106
393         394      Prof          A             31           19  Male  150564
394         395      Prof          A             42           25  Male  101738
395         396      Prof          A             25           15  Male   95329
396         397  AsstProf          A              8            4  Male   81035

Если нам нужно другое число строк (не 5, которые показываются по умолчанию), это легко исправить:

In [6]:
df.head(10)

Unnamed: 0       rank discipline  yrs.since.phd  yrs.service     sex  \
0           1       Prof          B             19           18    Male   
1           2       Prof          B             20           16    Male   
2           3   AsstProf          B              4            3    Male   
3           4       Prof          B             45           39    Male   
4           5       Prof          B             40           41    Male   
5           6  AssocProf          B              6            6    Male   
6           7       Prof          B             30           23    Male   
7           8       Prof          B             45           45    Male   
8           9       Prof          B             21           20    Male   
9          10       Prof          B             18           18  Female   

   salary  
0  139750  
1  173200  
2   79750  
3  115000  
4  141500  
5   97000  
6  175000  
7  147765  
8  119250  
9  129000

Выяснить размерность датафрейма можно с помощью атрибута `.shape`, в нем хранится кортеж с числом строк и числом столбцов:

In [7]:
print(df.shape)

(397, 7)


При желании можно извлечь только одно из чисел:

In [8]:
print(df.shape[0])
print(df.shape[1])

397
7


Техническое описание датафрейма можно получить, применив метод `.info()`:

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 397 entries, 0 to 396
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Unnamed: 0     397 non-null    int64 
 1   rank           397 non-null    object
 2   discipline     397 non-null    object
 3   yrs.since.phd  397 non-null    int64 
 4   yrs.service    397 non-null    int64 
 5   sex            397 non-null    object
 6   salary         397 non-null    int64 
dtypes: int64(4), object(3)
memory usage: 21.8+ KB


Какую информацию выдал метод `.info()`? Во-первых, он сообщил нам, что `df` является объектом `DataFrame`. Во-вторых, он вывел число строк (`RangeIndex: 397 entries`) и показал их индексы (`0 to 396`). В-третьих, он вывел число столбцов (`total 7 columns`). Наконец, он выдал информацию по каждому столбцу. Остановимся на этом поподробнее.

В выдаче выше представлено, сколько непустых элементов содержится в каждом столбце. Непустые элементы `non-null` – это все, кроме пропущенных значений, которые кодируются особым образом (`NaN` – от *Not A Number*). В нашей таблице все столбцы заполнены полностью. Далее указан тип каждого столбца, целочисленный `int64` и строковый `object`. Числа в конце означают объем памяти, который требуется для хранения, они зависят от битности системы.

Сводную статистическую информацию можно получить с помощью метода `.describe()`.

In [10]:
df.describe()

Unnamed: 0  yrs.since.phd  yrs.service         salary
count  397.000000     397.000000   397.000000     397.000000
mean   199.000000      22.314861    17.614610  113706.458438
std    114.748275      12.887003    13.006024   30289.038695
min      1.000000       1.000000     0.000000   57800.000000
25%    100.000000      12.000000     7.000000   91000.000000
50%    199.000000      21.000000    16.000000  107300.000000
75%    298.000000      32.000000    27.000000  134185.000000
max    397.000000      56.000000    60.000000  231545.000000

При применении ко всему датафрейму этот метод по умолчанию выбирает только числовые столбцы и выводит для них описательные статистики:

* `count` – число заполненных значений;
* `mean` – среднее арифметическое;
* `std` – стандартное отклонение (показатель разброса данных относительно среднего значения);
* `min` – минимальное значение;
* `max` – максимальное значение;
* `25%` – нижний квартиль (значение, которое 25% значений не превышают);
* `50%` – медиана (значение, которое 50% значений не превышают);
* `75%` – верхний квартиль (значение, которое 75% значений не превышают).

Если мы хотим описать только текстовые столбцы, нужно указать соответствующий тип внутри аргумента `include`:

In [11]:
df.describe(include = "object")

rank discipline   sex
count    397        397   397
unique     3          2     2
top     Prof          B  Male
freq     266        216   358

В таблице выше добавились новые строки `unique`, `top` и `freq`:

* `unique` – число уникальных значений в столбце (в `sex` их 2, `Male` и `Female`);
* `top` – мода, значение, которое встречается чаще всех (в `sex` больше значений `Male`, в выборке больше сотрудников-мужчин);
* `freq` – частота для значения в `top` (в `sex` 358, в выборке 358 сотрудников-мужчин).

Можно включить в описание все типы сразу, но это будет довольно громоздко (там, где характеристика неприменима, ставится пропуск `NaN`):

In [12]:
df.describe(include = "all")

Unnamed: 0  rank discipline  yrs.since.phd  yrs.service   sex  \
count   397.000000   397        397     397.000000   397.000000   397   
unique         NaN     3          2            NaN          NaN     2   
top            NaN  Prof          B            NaN          NaN  Male   
freq           NaN   266        216            NaN          NaN   358   
mean    199.000000   NaN        NaN      22.314861    17.614610   NaN   
std     114.748275   NaN        NaN      12.887003    13.006024   NaN   
min       1.000000   NaN        NaN       1.000000     0.000000   NaN   
25%     100.000000   NaN        NaN      12.000000     7.000000   NaN   
50%     199.000000   NaN        NaN      21.000000    16.000000   NaN   
75%     298.000000   NaN        NaN      32.000000    27.000000   NaN   
max     397.000000   NaN        NaN      56.000000    60.000000   NaN   

               salary  
count      397.000000  
unique            NaN  
top               NaN  
freq              NaN  
mean    113706.458438  
std      30289.038695  
min      57800.000000  
25%      91000.000000  
50%     107300.000000  
75%     134185.000000  
max     231545.000000

Если мы хотим запросить отдельно названия строк, нам пригодится знакомый атрибут `.index`:

In [13]:
print(df.index)

RangeIndex(start=0, stop=397, step=1)


А если столбцов – то атрибут `.columns`:

In [14]:
print(df.columns)

Index(['Unnamed: 0', 'rank', 'discipline', 'yrs.since.phd', 'yrs.service',
       'sex', 'salary'],
      dtype='object')
